pkg/engine/supervisor.go

```diff
@@ -0,0 +1,381 @@
+package engine
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/google/uuid"
+	"github.com/logrusorgru/aurora"
+	"github.com/otiai10/copy"
+	"github.com/testground/testground/pkg/api"
+	"github.com/testground/testground/pkg/config"
+	"github.com/testground/testground/pkg/logging"
+	"github.com/testground/testground/pkg/rpc"
+	"github.com/testground/testground/pkg/task"
+	"golang.org/x/sync/errgroup"
+	"path/filepath"
+	"strings"
+	"time"
+)
+
+type RunInput struct {
+	*api.RunRequest
+	Sources *api.UnpackedSources
+}
+
+type BuildInput struct {
+	*api.BuildRequest
+	Sources *api.UnpackedSources
+}
+
+func (e *Engine) startSupervisor(workers int) {
+	for i := 0; i < workers; i++ {
+		go e.worker(i)
+	}
+}
+
+func (e *Engine) worker(n int) {
+	logging.S().Infow("supervisor worker started", "worker_id", n)
+
+	var (
+		store = e.TaskStorage()
+		queue = e.TaskQueue()
+	)
+
+	for {
+		tsk, err := queue.Pop()
+		if err == task.ErrQueueEmpty {
+			time.Sleep(time.Second)
+			continue
+		}
+
+		ctx, _ := context.WithTimeout(context.Background(), time.Minute*30)
+
+		err = store.AppendTaskState(tsk.ID, task.StateProcessing)
+		if err != nil {
+			logging.S().Errorw("could not update task status", "err", err)
+		}
+		logging.S().Infow("worker processing task", "worker_id", n, "task_id", tsk.ID)
+
+		var data interface{}
+
+		switch tsk.Type {
+		case task.TaskRun:
+			data, err = e.doRun(ctx, tsk.ID, tsk.Input.(*RunInput))
+		case task.TaskBuild:
+			data, err = e.doBuild(ctx, tsk.Input.(*BuildInput))
+		default:
+			// wut
+		}
+
+		err = store.MarkCompleted(tsk.ID, err, data)
+		if err != nil {
+			logging.S().Errorw("could not update task status", "err", err)
+		}
+		logging.S().Infow("worker completed task", "worker_id", n, "task_id", tsk.ID)
+	}
+}
+
+func (e *Engine) doBuild(ctx context.Context, input *BuildInput) ([]*api.BuildOutput, error) {
+	sources := input.Sources
+	comp, err := input.Composition.PrepareForBuild(&input.Manifest)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := comp.ValidateForBuild(); err != nil {
+		return nil, fmt.Errorf("invalid composition: %w", err)
+	}
+
+	// TODO: how to pass this to the client with --wait? Should we?
```

Comment:
Yes, we should. We should also persist the execution trace for inspection with `testground logs`.